### DISCUSSION TITLE
[Proposal] Have a way to enforce field or property initialization on the constructor

### DISCUSSION THREAD

**Discussion by opcodewriter** [dotnet/csharplang#8142] [Language Ideas]
C# does not have a way to enforce having a property initialized on the constructor. 

The closest thing is using the _required_ keyword, but it works only if you use _object initializer_.

The following code will give a compilation error:

> Required member 'Program.Person.Id' must be set in the object initializer or attribute constructor.

```
    class Person
    {
        public required int Id { get; init; }

        public Person(int id) { Id = id; };
    }

var person = new Person(1);

```

Setting the [SetRequiredPropertiers] does not help, it basically tells the compiler to believe all the required properties are set.

I don't want to trick the compiler in any way, that defeats the whole purpose why I need this feature.

I'm looking for a solution where the compilers actually checks if the required fields or properties are actually set, like how _required_ works for the object initializers.

Such a feature be very handy, it really ensures the required fields / properties are set even when using a constructor.

I am NOT asking for the _required_ keyword to be modified, such a change won't be possible as it would break people's code.
I don't care if a new feature/keyword would be introduced.

**Comment by HaloFour** [comment]
You can use [SetRequiredPropertiesAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.setsrequiredmembersattribute?view=net-8.0) to specify that the constructor satisfies the required properties of the class:

```cs
    class Person
    {
        public required int Id { get; init; }

        [SetsRequiredProperties]
        public Person(int id) { Id = id; };
    }

var person = new Person(1);
```

The attribute is an all-or-nothing switch, meaning that a constructor adorned with it will be expected to have set all of the required properties.

**Reply by opcodewriter** [reply]
I edited my question. The attribute does not help, it tells the compiler "Blindly believe everything which is required is set". 

I don't want to trick the compiler do that, it defeats the whole purpose, which is to actually check if all required fields / properties are set.

**Reply by HaloFour** [reply]
The attribute exists for you to explicitly specify the contract expected by the class.  Having the developer explicitly specify that contract is intentional and important as it would be problematic if a tweak to the constructor body would break that public contract.  Also, constructor bodies are opaque to the compiler.  Even if the compiler could perform such a check on classes defined within your own project, it couldn't do that for classes you take as a reference.

Also, see: https://github.com/dotnet/csharplang/discussions/6405

**Reply by opcodewriter** [reply]
But the compiler already does that when using the _required_ specifier. All I want is the same thing when also using the c-tor, not just the object initializer

**Comment by colejohnson66** [comment]
Related: #8137

**Reply by opcodewriter** [reply]
Thanks but that issue is about making the attribute specify which fields are set by the c-tor. I am not looking for that. 

What I want is a way to enforce setting required properties, not only by the object initializer, but also by the c-tor.

For example, if I set a field as required and it's not set in any of the c-tors, the compiler should error.

**Reply by CyrusNajmabadi** [reply]
> For example, if I set a field as required and it's not set in any of the c-tors, the compiler should error.

that's super weird.  the point of a required-prop is to say that someone should be setting them by doing `new X() { YourProp = ... }`.  If you want the constructor to be required to assign it, then a required prop isn't the right thing.

**Reply by opcodewriter** [reply]
I'm not asking to extend the current `required` keyword! That would not be even possible, it would break a lot of people's code.

That's not the point. I don't care if a new modifier would be introduced.

**Reply by HaloFour** [reply]
This sounds like a case that would be best solved by an analyzer.  As a language feature I think there are too many cases where it is impractical for the compiler to track whether or not a certain member is initialized due to indirection through helper methods, etc.

**Reply by opcodewriter** [reply]
Why would it be impractical for the compiler to track it when using the c-tor, when it already does it when using object initializer?
What are you talking about?

**Reply by HaloFour** [reply]
Object initializer syntax is very limited in what it can do.  There is no branching or control statements, nor can you invoke helper methods which would be responsible for partial initialization.  Constructors can do all of these things, and it would be very impractical for the compiler to perform analysis across all such code to consider whether or not whatever the property was initialized.  That's even before you take into consideration the fact that you can't "initialize" whatever is the backing store of a property without calling the setter of that property.

For the simplest of scenarios, sure, enforcing required members in the constructor makes sense.  But there are too many cases that make it substantially more difficult.  The team already looked into this when evaluating whether or not the compiler would try to enforce `SetRequiredMembers` from within the constructor and found that it quickly spun out into edge cases.

**Reply by opcodewriter** [reply]
Thanks, I see what you're saying. It's unfortunate. What are the scenarios for the `SetRequiredMembers` attribute? I don't see how telling the compiler to believe you is helpful.

**Reply by HaloFour** [reply]
It's to declare to a consumer of the class that the constructor satisfies the required properties, so that they don't need to also include an initializer:

```cs
public class C {
    public required string Foo { get; init; }
    public required string Bar { get; init; }

    public C() { }

    [SetRequiredMembers]
    public C(string foo, string bar) {
        Foo = foo;
        Bar = bar;
    }
}

// when using parameterless constructor the initializer is required
var c1 = new C() {
    Foo = "foo",
    Bar = "bar"
};

// when using the constructor with parameters the initializer is not required
var c2 = new C("foo", "bar");
```

**Reply by opcodewriter** [reply]
Thanks, I know how it works, what I was trying to say is, you basically lie the compiler. I'm not seeing how such a feature is an useful option in the language. It does not help in the making sure you actually don't forget to set it.

**Reply by HaloFour** [reply]
The alternative is to not have that option at all, and any class with `required` members always requires that those members be set via initializers regardless of what constructor is used or what it does.
